pkg/storage/unified/resource/index_server.go

```diff
@@ -0,0 +1,138 @@
+package resource
+
+import (
+	"context"
+	"errors"
+	"strings"
+
+	"google.golang.org/grpc"
+)
+
+type indexServer struct {
+	s     *server
+	index *Index
+	ws    *indexWatchServer
+}
+
+func (is indexServer) Search(ctx context.Context, req *SearchRequest) (*SearchResponse, error) {
+	res := &SearchResponse{}
+	return res, nil
+}
+
+func (is indexServer) History(ctx context.Context, req *HistoryRequest) (*HistoryResponse, error) {
+	return nil, nil
+}
+
+func (is indexServer) Origin(ctx context.Context, req *OriginRequest) (*OriginResponse, error) {
+	return nil, nil
+}
+
+func (is *indexServer) Index(ctx context.Context, req *IndexRequest) (*IndexResponse, error) {
+	if req.Key == nil {
+		is.index = NewIndex(is.s, Opts{})
+		err := is.index.Init(ctx)
+		if err != nil {
+			return nil, err
+		}
+		return nil, nil
+	}
+
+	err := is.index.Index(ctx, &Data{Key: req.Key, Value: req.Value})
+	if err != nil {
+		return nil, err
+	}
+	return nil, nil
+}
+
+// Init sets the resource server on the index server
+// so we can call the resource server from the index server
+// TODO: a chicken and egg problem - index server needs the resource server but the resource server is created with the index server
+func (is *indexServer) Init(ctx context.Context, rs *server) error {
+	is.s = rs
+	is.ws = &indexWatchServer{
+		is:      is,
+		context: ctx,
+	}
+	// TODO: how to watch all resources?
```

Comment:
afaik you can only watch a single resource in k8s. But, if you go to `/apis` you can get all the api resources that we have and start watches on them all. I'm not sure if its guaranteed that every resource also lives in unified storage. @ryantxu would know more

Lets just start with playlists for this PR though 👍

Reply:
I think if you use the [resourceClient](https://github.com/grafana/grafana/blob/main/pkg/storage/unified/resource/client.go#L36) we should be able to list / watch any resource